File: android/chat-client/src/main/kotlin/com/anapaya/chat/client/ChatClient.kt
// Copyright 2026 Anapaya Systems

package com.anapaya.chat.client

import com.anapaya.chat.client.model.CreateRoomRequest
import com.anapaya.chat.client.model.LoginRequest
import com.anapaya.chat.client.model.LoginResponse
import com.anapaya.chat.client.model.Message
import com.anapaya.chat.client.model.MessagesResponse
import com.anapaya.chat.client.model.PostMessageRequest
import com.anapaya.chat.client.model.RegisterRequest
import com.anapaya.chat.client.model.Room
import com.anapaya.chat.client.model.RoomsResponse
import com.anapaya.chat.client.model.ServerInfo
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/** The chat API, over whatever [Transport] it is given. Holds the session token. */
public class ChatClient(private val transport: Transport) {
    private var token: String? = null

    /** Who is logged in. */
    public var username: String? = null
        private set

    public suspend fun health() {
        call("GET", "/healthz", authenticated = false)
    }

    public suspend fun serverInfo(): ServerInfo =
        decode(call("GET", "/server", authenticated = false))

    /** Creates the account. Does not log in: the server keeps the two apart. */
    public suspend fun register(username: String, password: String) {
        call(
            "POST",
            "/register",
            json = format.encodeToString(RegisterRequest(password = password, username = username)),
            authenticated = false,
        )
    }

    public suspend fun logIn(username: String, password: String) {
        val body = format.encodeToString(LoginRequest(password = password, username = username))
        val reply: LoginResponse = decode(call("POST", "/login", json = body, authenticated = false))

        token = reply.token
        this.username = username
    }

    public suspend fun rooms(): List<Room> = decode<RoomsResponse>(call("GET", "/rooms")).rooms

    public suspend fun createRoom(name: String): Room =
        decode(call("POST", "/rooms", json = format.encodeToString(CreateRoomRequest(name))))

    public suspend fun messagesNewest(room: Long, limit: Int = PAGE): List<Message> =
        decode<MessagesResponse>(call("GET", "/rooms/$room/messages?limit=$limit")).messages

    public suspend fun messagesAfter(room: Long, after: Long, limit: Int = PAGE): List<Message> =
        decode<MessagesResponse>(
            call("GET", "/rooms/$room/messages?after_seq=$after&limit=$limit"),
        ).messages

    public suspend fun send(room: Long, body: String) {
        call("POST", "/rooms/$room/messages", json = format.encodeToString(PostMessageRequest(body)))
    }

    public fun close(): Unit = transport.close()

    /** Sends, and turns anything that is not a success into a [ChatError]. */
    private suspend fun call(
        method: String,
        path: String,
        json: String? = null,
        authenticated: Boolean = true,
    ): String {
        val bearer = when {
            !authenticated -> null
            else -> token ?: throw ChatError.NotLoggedIn()
        }

        val reply = transport.send(ChatRequest(method, path, json, bearer))
        if (reply.status in 200..299) {
            return reply.body
        }

        val failure = refusal(reply.status, reply.body)
        // The token is gone, not refused, so retrying with it cannot work.
        if (failure is ChatError.SessionExpired) {
            token = null
            username = null
        }

        throw failure
    }

    private inline fun <reified T> decode(body: String): T =
        try {
            format.decodeFromString<T>(body)
        } catch (error: Exception) {
            throw ChatError.Protocol("a reply could not be decoded: ${error.message}")
        }

    private companion object {
        /** How many messages a page holds. */
        const val PAGE = 50

        val format = Json { ignoreUnknownKeys = true }

        fun refusal(status: Int, body: String): ChatError {
            val envelope = try {
                format.decodeFromString<ErrorEnvelope>(body)
            } catch (error: Exception) {
                return ChatError.Protocol("$status carried no error envelope: ${error.message}")
            }

            return when (envelope.error.code) {
                "expired_token", "unauthorized" -> ChatError.SessionExpired()
                else -> ChatError.Api(status, envelope.error.code, envelope.error.message)
            }
        }
    }
}

/** The body of every failing response. `code` is a `String`, so a code added later still decodes. */
@Serializable
private data class ErrorEnvelope(@SerialName("error") val error: ApiFailure)

@Serializable
private data class ApiFailure(
    @SerialName("code") val code: String,
    @SerialName("message") val message: String,
)

File: ios/ChatClient/Sources/ChatClient/ChatClient.swift
// Copyright 2026 Anapaya Systems

import Foundation

/// The chat API, over whatever ``Transport`` it is given. Holds the session token.
public actor ChatClient {
    /// How many messages a page holds.
    public static let page = 50

    private let transport: Transport
    private var token: String?

    /// Who is logged in.
    public private(set) var username: String?

    public init(transport: Transport) {
        self.transport = transport
    }

    public func health() async throws {
        _ = try await call("GET", "/healthz", authenticated: false)
    }

    public func serverInfo() async throws -> ServerInfo {
        try decode(await call("GET", "/server", authenticated: false))
    }

    /// Creates the account. Does not log in: the server keeps the two apart.
    public func register(username: String, password: String) async throws {
        let body = RegisterRequest(password: password, username: username)
        _ = try await call("POST", "/register", json: encode(body), authenticated: false)
    }

    public func logIn(username: String, password: String) async throws {
        let body = LoginRequest(password: password, username: username)
        let reply: LoginResponse = try decode(
            await call("POST", "/login", json: encode(body), authenticated: false))

        token = reply.token
        self.username = username
    }

    public func rooms() async throws -> [Room] {
        let reply: RoomsResponse = try decode(await call("GET", "/rooms"))
        return reply.rooms
    }

    public func createRoom(name: String) async throws -> Room {
        try decode(await call("POST", "/rooms", json: encode(CreateRoomRequest(name: name))))
    }

    public func messagesNewest(room: Int64, limit: Int = page) async throws -> [Message] {
        let reply: MessagesResponse = try decode(
            await call("GET", "/rooms/\(room)/messages?limit=\(limit)"))
        return reply.messages
    }

    public func messagesAfter(room: Int64, after: Int64, limit: Int = page) async throws -> [Message] {
        let reply: MessagesResponse = try decode(
            await call("GET", "/rooms/\(room)/messages?after_seq=\(after)&limit=\(limit)"))
        return reply.messages
    }

    public func send(room: Int64, body: String) async throws {
        _ = try await call(
            "POST", "/rooms/\(room)/messages", json: encode(PostMessageRequest(body: body)))
    }

    public func close() async {
        await transport.close()
    }

    /// Sends, and turns anything that is not a success into a ``ChatError``.
    private func call(
        _ method: String,
        _ path: String,
        json: String? = nil,
        authenticated: Bool = true
    ) async throws -> String {
        var bearer: String?
        if authenticated {
            guard let held = token else { throw ChatError.notLoggedIn }
            bearer = held
        }

        let reply = try await transport.send(
            ChatRequest(method: method, path: path, json: json, bearer: bearer))
        if (200...299).contains(reply.status) {
            return reply.body
        }

        let refusal = Self.refusal(status: reply.status, body: reply.body)
        // The token is gone, not refused, so retrying with it cannot work.
        if refusal == .sessionExpired {
            token = nil
            username = nil
        }

        throw refusal
    }

    private func encode<T: Encodable>(_ value: T) throws -> String {
        guard let json = try? JSONEncoder().encode(value),
              let text = String(data: json, encoding: .utf8)
        else {
            throw ChatError.protocolError("a request could not be encoded")
        }
        return text
    }

    private func decode<T: Decodable>(_ body: String) throws -> T {
        do {
            return try JSONDecoder().decode(T.self, from: Data(body.utf8))
        } catch {
            throw ChatError.protocolError("a reply could not be decoded: \(error)")
        }
    }

    private static func refusal(status: Int, body: String) -> ChatError {
        guard let envelope = try? JSONDecoder().decode(ErrorEnvelope.self, from: Data(body.utf8))
        else {
            return .protocolError("\(status) carried no error envelope")
        }

        switch envelope.error.code {
        case "expired_token", "unauthorized":
            return .sessionExpired
        default:
            return .api(
                status: status, code: envelope.error.code, message: envelope.error.message)
        }
    }
}

/// The body of every failing response. `code` is a `String`, so a code added later still decodes.
struct ErrorEnvelope: Decodable {
    let error: ApiFailure

    struct ApiFailure: Decodable {
        let code: String
        let message: String
    }
}
